pkg/pipeline/pipeline_prod.go

```diff
@@ -91,83 +102,142 @@ func newPipeline(input *InputBin, output *OutputBin) (*Pipeline, error) {
 		pipeline: pipeline,
 		output:   output,
 		removed:  make(map[string]bool),
-		started:  make(chan struct{}, 1),
+		started:  make(chan struct{}),
 		closed:   make(chan struct{}),
 	}, nil
 }
 
-func (p *Pipeline) Start() error {
-	loop := glib.NewMainLoop(glib.MainContextDefault(), false)
+func (p *Pipeline) Run() error {
+	// add watch
+	p.loop = glib.NewMainLoop(glib.MainContextDefault(), false)
 	p.pipeline.GetPipelineBus().AddWatch(func(msg *gst.Message) bool {
 		switch msg.Type() {
 		case gst.MessageEOS:
-			logger.Debugw("EOS received")
+			// EOS received - close and return
+			logger.Debugw("EOS received, stopping pipeline")
 			_ = p.pipeline.BlockSetState(gst.StateNull)
 			logger.Debugw("pipeline stopped")
-			loop.Quit()
+
+			p.loop.Quit()
 			return false
 		case gst.MessageError:
+			// handle error if possible, otherwise close and return
 			gErr := msg.ParseError()
-			handled := p.handleError(gErr)
-			if !handled {
-				loop.Quit()
+			err, handled := p.handleError(gErr)
+			if handled {
+				logger.Errorw("error handled", errors.New(gErr.Error()))
+			} else {
+				p.err = err
+				p.loop.Quit()
 				return false
 			}
-			logger.Debugw("handled error", "error", gErr.Error())
+		case gst.MessageStateChanged:
+			if msg.Source() == pipelineSource && p.startedAt.IsZero() {
+				_, newState := msg.ParseStateChanged()
+				if newState == gst.StatePlaying {
+					p.startedAt = time.Now()
+					close(p.started)
+				}
+			}
 		default:
 			logger.Debugw(msg.String())
 		}
+
 		return true
 	})
 
-	// start playing
-	err := p.pipeline.SetState(gst.StatePlaying)
-	if err != nil {
+	// set state to playing (this does not start the pipeline)
+	if err := p.pipeline.SetState(gst.StatePlaying); err != nil {
 		return err
 	}
 
-	// Block and iterate on the main loop
-	close(p.started)
-	loop.Run()
-	return nil
+	// run main loop
+	p.loop.Run()
+	return p.err
+}
+
+func (p *Pipeline) GetStartTime() time.Time {
+	select {
```

Comment:
what's the purpose of this? if it's not started, would it just block then?

Reply:
yeah, a goroutine is waiting to get the correct start time (recorder.go#L95) since it takes about 5 seconds for it to actually start recording